Code/Tools/FontImporter/TextureAtlasLayout.cpp

```diff
@@ -0,0 +1,195 @@
+    }
+    else
+    {
+      left = EZ_DEFAULT_NEW(ezTextureAtlasNode, node->x, node->y, node->Width, element.Input.Height);
+      right = EZ_DEFAULT_NEW(ezTextureAtlasNode, node->x, node->y + element.Input.Height, node->Width, node->Height - element.Input.Height);
+    }
+
+    node->Left = std::move(left);
+
+    node->Right = std::move(right);
+
+    return AddToNode(node->Left.Borrow(), element, allowGrowth);
+  }
+}
+
+ezDynamicArray<ezTextureAtlasPage> ezTextureAtlasUtility::CreateTextureAtlasLayout(ezDynamicArray<ezTextureAtlasElement>& elements, ezUInt32 width, ezUInt32 height, ezUInt32 maxWidth, ezUInt32 maxHeight, bool powerOfTwo)
+{
+  // Preserve original index before sorting
+  for (ezUInt32 i = 0; i < elements.GetCount(); i++)
+  {
+    elements[i].Output.Index = i;
+    elements[i].Output.Page = -1;
+  }
+
+  elements.Sort([](const ezTextureAtlasElement& left, const ezTextureAtlasElement& right) -> bool {
+    return left.Input.Width * left.Input.Height > right.Input.Width * right.Input.Height;
+  });
+
+  ezDynamicArray<ezTextureAtlasLayout> layouts;
+  ezUInt32 remainingCount = elements.GetCount();
+
+  while (remainingCount > 0)
+  {
+    layouts.PushBack(ezTextureAtlasLayout(width, height, maxWidth, maxHeight, powerOfTwo));
+
+    ezTextureAtlasLayout& currentLayout = layouts.PeekBack();
+
+    // Find largest unassigned element that fits
+    ezUInt32 sizeLimit = ezMath::MaxValue<ezUInt32>();
+
+    while (true)
+    {
+      ezUInt32 largestId = -1;
+
+      // Assumes elements are sorted from largest to smallest
+      for (ezUInt32 i = 0; i < elements.GetCount(); i++)
+      {
+        if (elements[i].Output.Page == -1)
+        {
+          ezUInt32 size = elements[i].Input.Width * elements[i].Input.Height;
+
+          if (size < sizeLimit)
+          {
+            largestId = i;
+            break;
+          }
+        }
+      }
+
+      if (largestId == -1)
+      {
+        // Nothing fits, start a new page
+        break;
+      }
+
+      ezTextureAtlasElement& element = elements[largestId];
+
+      // Check if an element is too large to ever fit
+      if (element.Input.Width > maxWidth || element.Input.Height > maxHeight)
+      {
+        ezLog::Warning("Some of the provided elements don't fit in an atlas of provided size. Returning empty array of pages.");
+
+        return ezDynamicArray<ezTextureAtlasPage>();
+      }
+
+      if (currentLayout.AddElement(element))
+      {
+        element.Output.Page = layouts.GetCount() - 1;
+        remainingCount--;
+      }
+      else
+      {
+        sizeLimit = element.Input.Width * element.Input.Height;
+      }
+    }
+  }
+
+  ezDynamicArray<ezTextureAtlasPage> pages;
+
+  pages.Reserve(layouts.GetCount());
+
+  for (auto& layout : layouts)
+  {
+    pages.PushBack({layout.GetWidth(), layout.GetHeight()});
+  }
+
+  return pages;
```

Comment:
Again, there is only so much that move semantics can do to prevent unnecessary work. That copying containers around is 'efficient' is unfortunately a myth that the C++ standards committee is actively pushing, for reasons that are not quite clear to me (I think they want to believe this themselves). The reality is, that when you move a container, there are often things that cannot be optimized away, but without digging really deep into your data structures (and ours), it is hard to tell whether it will be fast or slow.

Consequently, we just never do this. Instead we would pass in the output container as a ref parameter, and work directly on it. Some people consider that pattern ugly, but talking about beauty and using C++ at the same time is weird anyway ;-)